### PR TITLE
semantic_text: Add index metadata information for inference field mappers

### PR DESCRIPTION
**Pull request by carlosdelest**
As prior work for `semantic_text` new field type, these changes add new `IndexMetadata` information for field types that need inference to be performed.

As inference is done in the coordinator node to avoid re-doing it at shard / replica level, the coordinator needs to check for the need to perform inference for specific fields in an index.

Given that the coordinator node does not necessarily have mapping information for all indices (only for those that have shards in the node), the field inference information must be stored in the `IndexMetadata` and broadcasted to all nodes.

This change adds:
- An `InferenceFieldMetadata `class that contains the information for fields that need inference
- - A new `InferenceFieldMapper` interface that will be implemented by `semantic_text` and any other field types that require inference, so this information can be retrieved from the field mapping
- Changes to the mapping lookup and mapping service creation to include this information from the mapping into the `IndexMetadata`

### CODE REVIEW

**Review by jimczi — APPROVED** [review]
Few minor comments, LGTM otherwise. Let's remove the draft status and have another reviewer since the targeted branch is main?

File: server/src/main/java/org/elasticsearch/index/mapper/InferenceFieldMapper.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.elasticsearch.cluster.metadata.InferenceFieldMetadata;
+import org.elasticsearch.inference.InferenceService;
+
+import java.util.Set;
+
+/**
+ * Field mapper that requires to transform its input before indexation through the {@link InferenceService}.
+ */
+public interface InferenceFieldMapper {
+    String NAME = "_inference";
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/InferenceFieldMapper.java]: Done, 7a2b70b9c33b064feedbce2aa5b7131b9f8a8b83

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 2]
```diff
@@ -540,6 +540,8 @@ public Iterator<Setting<?>> settings() {
 
     public static final String KEY_SHARD_SIZE_FORECAST = "shard_size_forecast";
 
+    public static final String KEY_FIELD_INFERENCE = "field_inference";
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java]: Done, 7a2b70b9c33b064feedbce2aa5b7131b9f8a8b83

File: server/src/main/java/org/elasticsearch/TransportVersions.java [hunk 3]
```diff
@@ -163,6 +163,7 @@ static TransportVersion def(int id) {
     public static final TransportVersion CCR_STATS_API_TIMEOUT_PARAM = def(8_622_00_0);
     public static final TransportVersion ESQL_ORDINAL_BLOCK = def(8_623_00_0);
     public static final TransportVersion ML_INFERENCE_COHERE_RERANK = def(8_624_00_0);
+    public static final TransportVersion SEMANTIC_TEXT_FIELD_ADDED = def(8_625_00_0);
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: Agreed, 7a2b70b9c33b064feedbce2aa5b7131b9f8a8b83

File: server/src/test/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadataTests.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    protected boolean supportsUnknownFields() {
+        return true;
+    }
+
+    private static InferenceFieldMetadata createTestItem() {
+        String name = randomAlphaOfLengthBetween(3, 10);
+        String inferenceId = randomIdentifier();
+        String[] inputFields = generateRandomStringArray(5, 10, false, false);
+        return new InferenceFieldMetadata(name, inferenceId, inputFields);
+    }
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadataTests.java:65]: Could we have some tests covering the `null` cases that throw?

**Review by benwtrent — APPROVED** [review]
A minor concern. I think it all looks OK. My other concerns were about `null` being accidentally provided in the ctor as we have a ton of stuff that relies on the inference fields not being `null`, but instead `empty`, but it seems like we have adequate coverage and protection there. Just being paranoid :)

File: server/src/test/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadataTests.java (hunk 4 above)
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadataTests.java:65]: Sure! dc46e88381eac36eb62da901729c1acd1eb376af

**Comment by carlosdelest** [comment]
@elasticsearchmachine update branch

**Review by Mikep86 — APPROVED** [review]
LGTM, only two small qeustions/comments

File: server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        return builder.endObject();
+    }
+
+    public static InferenceFieldMetadata fromXContent(XContentParser parser) throws IOException {
+        final String name = parser.currentName();
+
+        XContentParser.Token token = parser.nextToken();
+        if (token == null) {
+            // no data...
+            return null;
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java:106]: It's not really possible as this method is called when parser.nextToken() is not null. I'll use a non null requirement to fail early in this case 👍

**Review by original-brownbear — APPROVED** [review]
LGTM couldn't come up with anything else either as discussed on Slack :) Thanks Carlos!

File: server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mikep86) [reply server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java]: At this point in the parsing, we're expecting the current token to be `START_OBJECT`, followed by one or more key-value pairs, then an `END_OBJECT`, correct?

If so, maybe it makes sense to check that the current token is `START_OBJECT` here and then error out on any `null`s returned by `nextToken` in the `while` loop (if the parser calls don't already handle that)

**Pull request merged**